CATL-1021: Display Webforms With Cases In Case View Actions Menu

vinuvarshith opened the pull request:
Overview
----------
As a user of CiviCase, I should be able to access webforms on a case so that I can update multiple fields in one go for a case.

Before
-------
Webforms are not accessible from within the case - however - you can add a case to webforms.

After
-----
Webforms which have a cases component added to them need to appear in the case menu ie.
forms where No of cases (in CiviCRM tab of the webform) =>1 should appear on the case actions dropdown.
I have added this as a submenu that pops out on hovering 'Webforms' menu item in civicase action menu.

jamienovick commented:
@guanhuan can you check what's going on here. We can't have all forms which have a case on them automatically showing in the menu can we?

shitijg commented:
@jamienovick @guanhuan - this is to show all the referral forms in the case actions menu. Do you envisage any other logic to show them here?

jamienovick commented:
a) This extension must be able to be deployed in CiviCRM with other CMS, and also when webform or webform CiviCRM are not installed on current implementation (acceptance criteria). IF we create this dependency on webform it must still work without it.

b) The problem is that there may be case webforms that you don't want to link in the menu on a case as they are not for the admin. Whilst I agree this is better then hardcoding the links, I'm not sure if there maybe something better we can do so admin can select which webforms are linked from a case? I would suggest an admin screen but that seems like a lot of work... @guanhuan @vinuvarshith any ideas?

c) Also are we passing the case ID in the URL to these forms in the link so the case that you are in the source?

vinuvarshith commented:
@jamienovick @shitijg @guanhuan 
a) We can have a check to see if the site uses Drupal and civicrm webforms and then execute.

b) Can we have a checkbox in the webform cases tab that says "Add a link to this webform in cases menu"? 
Or we could also have a separate admin page that lists all webforms and the admin could select which ones they want to be displayed in the case menu. The first would be a bit simpler to implement.

c) As of now this was not done. But this could be added with minor code change.

Once we confirm, I will add these features to this PR.

jamienovick commented:
**a) We can have a check to see if the site uses Drupal and civicrm webforms and then execute.**

Thanks please do.

**b) Can we have a checkbox in the webform cases tab that says "Add a link to this webform in cases menu"?**

That sounds like a workable solution, but would the code for that need to be a PR to webform_civicrm core or could it live in the new cases extension? Is it bad to have it live in the cases extension @omarabuhussein @MiyaNoctem ?

**Or we could also have a separate admin page that lists all webforms and the admin could select which ones they want to be displayed in the case menu. The first would be a bit simpler to implement.**

Agree, I prefer the first one also.

**c) As of now this was not done. But this could be added with minor code change.**

Please add this - thats an important part of the menu. The form should open in the context that this is the case that is being updated.

Once we confirm, I will add these features to this PR.

Merge state

MiyaNoctem commented:
@jamienovick @vinuvarshith I think it would certainly be better to have all this functionality within the extension, even if creating the Admin form requires a little more work. 

UX-wise it might be better to have the checkbox when creating the webform, but my guess is this will only need to be enabled for a few case types under specific use cases, so the trade-off of having a separate form to manage this might be a less of a problem than adding a dependency to a specific patch of the webforms module.

vinuvarshith commented:
@MiyaNoctem 
I have added an admin page where we can configure which webform link should be shown in the menu.

MiyaNoctem commented on css/caseView.css:
```diff
@@ -542,3 +542,19 @@
+  visibility: visible;
+  opacity: 1;
+  display: block;
+}
```
Missing new line at EOF.

MiyaNoctem commented on api/v3/Case/Getwebforms.php:
```diff
@@ -0,0 +1,45 @@
+  while ($dao->fetch()) {
+    $data = unserialize($dao->data);
+    if ($data['case']['number_of_case'] >= 0) {
+      $webforms[] = array('nid' => $dao->nid, 'title' => $dao->title, 'path' => drupal_get_path_alias('node/'.$dao->nid));
```
Please clean-up by adding new lines for each array element. Also, keep it consistent with either [] or array() syntax.

```php
$webforms[] = array(
  'nid' => $dao->nid, 
  'title' => $dao->title, 
  'path' => drupal_get_path_alias('node/'.$dao->nid)
);
```

MiyaNoctem commented on api/v3/Case/Getwebforms.php:
```diff
@@ -0,0 +1,45 @@
+ * Search for webforms that have a atleast 1 case attached to it.
+ *
+ * @param array $params
+ * @return array API result
```
Please add an empty line to separate @param from @return and @throws.

Eg.
```php
/**
 * ...
 *
 * @param array $params
 *
 * @return array API result
 *
 * @throws API_Exception
 */
```

MiyaNoctem commented on CRM/Civicase/Form/CaseWebforms.php:
```diff
@@ -0,0 +1,80 @@
+    $this->assign('nids', $webformids);
+    $this->assign('error', $errorMsg);
+
+    $this->addButtons([
```
You're using here short-hand array syntax, while most of the rest of the code you added uses `array()` syntax to define arrays. Which one should it be?

vinuvarshith commented:
@MiyaNoctem I have addressed your review comments. I have also added some updates. Pls check them too.

@jamienovick 
I wanted to confirm a few points with you
- I have added a separate admin page that will display all webforms that have a case component added to it. We can then 'check' which once we want to display in the cases menu.
- I have added conditions so that if 'Drupal' is not used or 'webform_civicrm' is not used, a warning message will be shown in the above admin page and will not break in the other CMSs.
- Based on a discussion I had with @shitijg, I have also added the case client id (contact id) to the webform url when redirecting.

jamienovick commented:
Thanks @vinuvarshith 

- I have added a separate admin page that will display all webforms that have a case component added to it. We can then 'check' which once we want to display in the cases menu.

Sounds good.

- I have added conditions so that if 'Drupal' is not used or 'webform_civicrm' is not used, a warning message will be shown in the above admin page and will not break in the other CMSs.

I think for other CMS perhaps we should just hide the page completely? For Drupal, yes we should state that webform_civicrm needs to be enabled.

- Based on a discussion I had with @shitijg, I have also added the case client id (contact id) to the webform url when redirecting.

Interesting. Not sure if we need it, but assume we do. What if we have multiple case clients @shitijg ?

shitijg commented:
@jamienovick we can't have a dynamic number of contacts on the webform, so I guess it will have to be the first / randomize()? @vinuvarshith

jamienovick commented:
So I'm questioning whether we should have a link with the contact in it at all. Why would we do that? Surely the form can dictate loading the appropriate contacts?

shitijg commented:
@vinuvarshith can you resolve the conflicts and see the changes requested by Camilo?

shitijg commented:
@vinuvarshith once ready with the changes suggested by Coleman, please PR to the new case extension and then deploy it on http://develop.civicase.ccuptest.co.uk. We have compubot configured in the newcivicase room. Contact @deb1990  for any questions. You also need to make sure that webform civicrm is installed and enabled there.

@vinuvarshith  please also feel free to ask @deb1990  for any styling help needed for the case actions dropdown

Please test the following as minimum (feel free to add more scenarios to test):

- If a form has cases enabled, it should show here: civicrm/case/webforms

- The form should not be available anymore if cases is disabled (negative test)

- Once the form checkbox is ticked here - civicrm/case/webforms, it should show in the case actions dropdown under a menu item “Webforms” which should further cascade to show a dropdown of all the webforms which have their checkbox ticked here: civicrm/case/webforms

- The webforms menu item should not show in caselist bulk actions (civicrm/case/a/#/case/list)

- Once a form is triggered, the URL should automatically have the CID of the “client” ie if Contact 2 is set as the client in the cases tab of the webform, when a form is triggered, it will have the case id set to the current case ID and CID2 as the contact ID of the client (it will be CID2 as the client is set to Contact 2 in the webform configurations)

- Form opens in a new window. All case fields that have values should be preloaded.

Thanks,
Shitij
